### ISSUE TITLE
Class difficulties are saved as strings

### ISSUE THREAD

**Issue by willgearty** [learning-unlimited/ESP-Website#3708]
The options for a class's difficulty is set by the global "teacherreg_difficulty_choices" tag. This tag is basically an ordered dictionary consisting of rankings/abbreviations (keys) and their descriptions (values). During class creation/editing, only the descriptions are shown, but the chosen key is saved as a string. This key is then shown on the course catalog and during student registration. However, this presents several problems:

1. The difficulty description is never shown outside of teacher registration (as far as I can tell). This means that students can only infer what the difficulty is based on the abbreviation. The default for this is some number of stars, but it's unclear to students how many stars is the max, etc.
2. When the choices tag is changed, this can possibly invalidate all course registration forms because the old chosen strings are no longer valid options. This similarly causes the difficulty filter on the course catalog to break (because the new options that are populated into the filter dropdown don't match the strings in the catalog).

I'm not too sure what the fix for this should be. Here are some possible options:

- We could have "difficulty option" model objects like we have flag categories which can be added and removed to individual programs? They would need to have an order such that they are listed in order in the course reg form and catalog filter dropdowns. But how do we backport this behavior to existing difficulty settings for existing classes? We could find all unique difficulty strings and add those as objects in a migration?
- We could force the tag to just be a list, but then we save the list index for the difficulty field. When we render the course in the catalog, we use the value at that list index. Unfortunately, this doesn't completely solve the tag change issues because the tag length could be shortened, causing the same problems. This would also have the problem related to backporting.
- Maybe we are fine with the current behavior regarding invalidating the reg form, and all we really need to fix is the catalog filter? Instead of using the choices tag, we could pull all unique difficulties from the classes in the catalog (although the entire page is cached which could be an issue), then just sort them alphanumerically? Although this still presents the issue where descriptions are never shown to students.

**Comment by willgearty** [comment]
@kkbrum brought this up and probably has thoughts.

**Comment by SuyashJain17** [comment]
Hi, I’ve been reviewing this and I agree there are two distinct problems here:

1. UX issue – Students only see the difficulty abbreviation (e.g., stars), not the description, which makes interpretation unclear.

2. Data integrity issue – Since difficulty values are stored as strings tied to a global tag, changes to the tag can invalidate existing class data and break catalog filtering.

Rather than modifying the tag structure directly (which risks breaking existing registrations), I’d suggest we approach this in stages:

- First, improve student visibility by displaying the difficulty description alongside the abbreviation in the catalog and registration pages.

- Then, evaluate whether difficulty should be modeled as a first-class entity (e.g., a DifficultyOption model scoped per program with ordering), similar to flag categories.

If we go the model route, we could write a migration that:

- Extracts all unique existing difficulty strings

- Creates corresponding model entries

- Maps existing classes to those entries

Before implementing anything large, I’d be happy to draft a short proposal comparing:

- Keeping the tag but improving rendering

- Moving to a model-backed solution

- Hybrid approaches

Let me know how you'd prefer to proceed, and I can take this forward.

**Comment by willgearty** [comment]
Yeah, this one's definitely tricky. Your strategy sounds like a good one, starting out with just improving visibility. Happy to give feedback on a proposal before moving forward.

**Comment by SuyashJain17** [comment]
Thanks! I took another look at this and tried to separate the UX problem from the data model problem.

For a first step, I’d suggest we improve visibility without changing how difficulty is stored. Right now the abbreviation is what students see, and the description is effectively hidden outside teacher registration. We could update the catalog and registration pages to display both the abbreviation and its full description (pulled from teacherreg_difficulty_choices). We could also adjust the filter dropdown to make it clearer what each option represents.

This would address the immediate student confusion without introducing migrations or risking inconsistencies with existing registrations.

If that feels like the right direction, we could then evaluate a structural change separately - for example, introducing a DifficultyOption model (possibly scoped per program) and migrating existing difficulty strings to it. That would give us better long-term stability if the global tag changes, but it’s a bigger change and probably worth treating as its own phase.

Let me know if you’d prefer starting with the smaller UX improvement, and I can draft a PR focused just on that.

**Comment by willgearty** [comment]
That sounds like a good start to me (the catalog and registration page changes). Thanks!
